docs/internals/layout_in_storage.rst

```diff
@@ -53,30 +64,57 @@ Mappings and Dynamic Arrays
+"in between" the state variables preceding and following them.
+Instead, they are considered to occupy only 32 bytes with regards to the
+:ref:`rules above <storage-inplace-encoding>` and the actual data of mappings and
+dynamcically-sized arrays is stored starting at a different
+storage slot that is computed using a Keccak-256 hash.
 
-The mapping or the dynamic array itself occupies a slot in storage at some position ``p``
-according to the above rule (or by recursively applying this rule for
-mappings of mappings or arrays of arrays). For dynamic arrays,
+Assume the storage location of the mapping or array ends up being a slot ``p``
+after applying :ref:`the storage layout rules <storage-inplace-encoding>`.
+For dynamic arrays,
 this slot stores the number of elements in the array (byte arrays and
 strings are an exception, see :ref:`below <bytes-and-string>`).
-For mappings, the slot is unused (but it is needed so that two equal mappings after each other will use a different
-hash distribution). Array data is located at ``keccak256(p)`` and the value corresponding to a mapping key
-``k`` is located at ``keccak256(k . p)`` where ``.`` is concatenation. If the value is again a
-non-elementary type, the positions are found by adding an offset of ``keccak256(k . p)``.
-
-So for the following contract snippet
-the position of ``data[4][9].b`` is at ``keccak256(uint256(9) . keccak256(uint256(4) . uint256(1))) + 1``::
-
+For mappings, the slot stays empty, but it is needed so that two mappings after each other will use a
+different hash distribution to store their data due to a different starting slot.
+
+Array data is located starting at ``keccak256(p)`` and it is laid out in the same way as
+statically-sized array data would: One element after the other, potentially sharing
+storage slots if the elements are not longer than 16 bytes. Dynamic arrays of dynamic arrays apply this
+rule recursively. The location of ``x[i][j]`` for ``x`` being ``uint32[][]`` is
+computed as follows (again, assuming ``x`` itself is stored at slot ``p``):
+The slot is ``keccak256(keccak256(p) + i) + floor(j / 8)`` (8 items per slot because 256/32 = 8) and
+the element can be obtained from the slot data ``v`` using ``(v >> ((j % 8) * 32)) & 0xffffffff``.
+
+The value corresponding to a mapping key ``k`` is located at ``keccak256(h(k) . p)``
+where ``.`` is concatenation and ``h`` is a function that is applied to the key depending on its type:
+
+- for value types, ``h`` pads the value to 32 bytes according to its alignment
+- for strings and byte arrays, ``h`` computes the ``keccak256`` hash of the unpadded data.
+
+If the mapping value is a
+non-value type, the location is the starting slot of the data. If the value is of struct type,
+for example, you have to add an offset corresponding the the struct member to reach the member.
+
+Let us compute the storage location of ``data[4][9].c`` in the following contract.
```

Comment:
Maybe it's nicer to say
```
As an example, consider the following contract:

<Contract>

Let us compute the storage location of...
```

otherwise one needs to look/scroll down and then move back to understand this... but maybe fine...